llvm/lib/Transforms/Yk/ModuleClone.cpp

```diff
@@ -70,13 +70,56 @@ struct YkModuleClone : public ModulePass {
     }
   }
 
+  /**
+   * This function iterates over all functions in the `FinalModule`.
+   * If cloned function calls are identified within the original function
+   * instructions, they are redirected to the original function instead.
+   *
+   * **Example Scenario (before calling `updateFunctionCalls` function):**
+   * - Function `f` calls function `g`.
+   * - Function `g` is cloned as `__yk_clone_g` and all its callsites
+   * are updated appropriately.
+   * - Function `f` is not cloned because its address is taken.
+   * - As a result, function `f` calls `__yk_clone_g` instead of `g`.
+   *
+   * **Reasoning:**
+   * In `YkIRWriter` we only serialise non-cloned functions. We want
```

Comment:
AFAIC this isn't really related to the serialiser. The real reason we want unoptimised functions to only call other unoptimised functions is because otherwise block recording won't work. In other words, if `f` calls `g_opt`, then `f` will record the blocks, but `g_opt` won't, we get incorrect traces.

Reply:
youre right.